### ISSUE TITLE
CSS in Imba way

### ISSUE THREAD

**Issue by altteo** [imba/imba#303]
Maybe we could include css in Imba as it was done with html?
I see it this way: we can have additional to render function (for example style) where we write css for the current Imba tag with readable and light syntax (stylus alike) and with hierarchy.
At the moment css in the comments looks at the wrong place.
‘’’
def style
     **self**
           color:red
           display: block
           **div**
                  color: white
‘’’

**Comment by somebee** [comment]
It is interesting to explore this. I think we should keep allowing people to use raw css for a while, but I've been playing around with some ways to do css myself. My current preference is something like:

```imba
tag hello-there

	css .disabled
		display: block
		font-style: italic

	css ul li
		background: white

		&:hover
			background: whitesmoke

	def render
		<self>
			<ul> for item in items
				<li> item
```

Suggestion is to add a `css` keyword like `get/set/def` in tag class bodies

**Comment by altteo** [comment]
Yep. Pretty good. But the killer feature will be if clause.

```
tag Comp
    @done=false

    css .default
      color:red if done
```

**Comment by somebee** [comment]
Yeah, that is another level. Difficult to implement in a performant way, but I'm sure we could find a way :)

**Comment by altteo** [comment]
That will force people to switch from raw css. This is needed almost anywhere and is far more important then slots. You traveled a long way to just stop at the very end! I believe in you! :-)

**Comment by altteo** [comment]
I thought about it for a while... As I understand we do not consider css-in-js and inline styling, so compiler should generate a css file. If statement will take away a lot of current hand-made mess with different classes like `<div .done=@done .active=!@done)>`. So compiler should generate different classes and connect them to the tag with initial class.
`<div .default>` should become `<div .default .default_done=@done .default_notdone=!@done>`
I mean we should compile class with “clear” css properties (without conditions) and for every “if” clause we should compile two styles (containing only conditional css property) and set them to the tag as usual.

**Comment by iamtirado** [comment]
This is a nice implementation idea. I like it.

**Comment by iamtirado** [comment]
```imba
tag hello-there
    css scoped
		.disabled
        	display: block
        	font-style: italic
    	ul li
        	background: white
        	&:hover
            	background: whitesmoke
	css 
		body
			background-color: #efefef
    def render
        <self>
            <ul> for item in items
                <li> item
```

**Comment by iamtirado** [comment]
I agree with altteo. It should go to a css document, rather than css-in-js.

**Comment by iamtirado** [comment]
Is this syntax more realistic?

```imba
let theme =
    primary: 'tomato'
def color val
    "color: {val}"
tag app-root
    <self>
        <h1> "Hello World!"
    <style css scoped>
        h1 {
            color(theme.primary)
        }
```

**Comment by iamtirado** [comment]
@somebee 
how about having it all under a single "css" keyword
```
css
  .disabled
    display: block
    font-style: italic
  ul li 
    ...
```

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by somebee** [comment]
Adding custom syntax for css is something I'm very interested in, but likely not for 2.0. Maybe in a couple of months :) Just need to stabilize and get 2.0 production ready first!

**Comment by ephetic** [comment]
In the same way imba is making HTML/JS domain tasks easy (drop noisy closing tags, register event handlers with common modifiers, etc), what problems in the CSS domain are trying to improve with imba-css?  I'd love if we could come up not just a pleasing syntax, but a layer of improvement over CSS.

**Comment by somebee** [comment]
There are many exciting things we could improve. Like @altteo has proposed here earlier we could allow conditional styling that is automatically compiled to underlying classes (added/removed automatically). We could also couple css variables and dom properties in a more natural way. Also, for responsive design you can imagine doing something like

```imba
tag app-item
	css
		if offsetWidth < 500
			display: flex
			font-size: 14px
		if offsetWidth < 300
			flex-direction: row
	...
```

where we automatically map these size-queries to classes that are then set using ResizeObserver etc. There are tons of cool things we can do, But first, we need to ship version 2 :D

**Comment by scanf** [comment]
Agree with strongly with the last comment. We should defer everything related to styling to post v2 unless someone can make a really compelling argument to delay v2 for these kinds of changes.

**Comment by iamtirado** [comment]
@scanf How about V2.1?

**Comment by scanf** [comment]
At the end of the day it's up to @somebee but IMO the right thing is to defer this to v3 and come back to it on 1-2 years.
